### PR TITLE
CIRC-238 Respect tenant-selected timezone for due date calculations

### PR DESCRIPTION
**Pull request by OleksiiMaksymov**
### Description
Load tenant-selected time-zone from mod-config and apply it for closed library due date management calculations

### Approach
1. add additional request to `mod-configuration`
2. get a date time zone and apply to strategies

### Link
[CIRC-238](https://issues.folio.org/browse/CIRC-238)

### CODE REVIEW

File: src/test/java/api/support/fakes/FakeOkapi.java [hunk 1]
```diff
@@ -329,6 +332,19 @@ private void registerLibraryHours(Router router) {
       });
   }
 
+  private void registerConfiguration(Router router) {
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/fakes/FakeOkapi.java]: Is there a reason that the fake configuration is using this mechanism rather than the standard storage interface fake?

It is my understanding that the calendar calculate openings endpoint needs this because of the way the calculations are canned responses (and not dependent upon the specified opening times themselves).

That limitation meant that following the closed due date changes, it is now not possible to use the circulation API tests with the real module instances. This diminished a significant source of integration feedback. I'd prefer we didn't add to this unless it is definitely necessary.

Please let me know if this doesn't make sense and I can try to explain further.

File: src/test/java/api/support/APITests.java [hunk 2]
```diff
@@ -188,7 +191,6 @@ public void beforeEach()
       useDefaultRollingPolicyCirculationRules();
     }
 
-    configureDefaultTimezone();
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/APITests.java]: Can you help me understand why this was removed?

I'm aware that the default is UTC, and so this is technically redundant, I'd prefer it was explicitly set.

(I'd also prefer that the system fail if this wasn't configured, but I've accepted that is out of scope for this work at the moment)

Comment (OleksiiMaksymov) [reply src/test/java/api/support/APITests.java]: yes, right, by default we have UTC default time zone. 
I used these settings in tests for the default time zone, custom, and UTC values

File: src/main/java/org/folio/circulation/support/JsonPropertyWriter.java [hunk 3]
```diff
@@ -54,7 +54,7 @@ public static void write(
 
   public static void write(JsonObject to, String propertyName, DateTime value) {
     if(value != null) {
-      write(to, propertyName, value.toString(ISODateTimeFormat.dateTime()));
+      write(to, propertyName, value.toString(ISODateTimeFormat.dateTime().withZoneUTC()));
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/support/JsonPropertyWriter.java]: This change has been made to ensure that all date times are converted to UTC when written to JSON?

File: descriptors/ModuleDescriptor-template.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
   "provides": [
     {
       "id": "circulation",
-      "version": "7.4",
+      "version": "7.5",
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: Why has the circulation interface version changed?

There does not appear to be any changes to the interface definition?

File: descriptors/ModuleDescriptor-template.json [hunk 5]
```diff
@@ -156,7 +157,8 @@
             "circulation-storage.loan-policies.item.get",
             "circulation-storage.request-policies.item.get",
             "circulation-storage.fixed-due-date-schedules.item.get",
-            "circulation-storage.fixed-due-date-schedules.collection.get"
+            "circulation-storage.fixed-due-date-schedules.collection.get",
+            "configuration.entries.collection.get"
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: Please add a dependency defined for the `configuration` interface

File: src/main/java/org/folio/circulation/domain/LoanRenewalService.java [hunk 6]
```diff
@@ -27,7 +27,7 @@ private LoanRenewalService(LoanPolicyRepository loanPolicyRepository) {
   }
 
   public CompletableFuture<HttpResult<Loan>> overrideRenewal(Loan loan, DateTime dueDate, String comment) {
-    return loanPolicyRepository.lookupLoanPolicy(loan)
+    return loanPolicyRepository.lookupLoanPolicy(loan, DateTimeZone.UTC)
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/LoanRenewalService.java]: Why is UTC hardcoded here, is it because timezone is not yet supported for overridden renewal?

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java [hunk 7]
```diff
@@ -17,18 +17,30 @@
...
+    }
+  }
+
+  // for test only
+  static FixedDueDateSchedules from(JsonObject representation) {
+    if (representation == null) {
+      return new NoFixedDueDateSchedules();
+    } else {
+      return new FixedDueDateSchedules(JsonArrayHelper.toList(
+        representation.getJsonArray("schedules")), DateTimeZone.UTC);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: Similarly, why is UTC hardcoded here?

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java [hunk 8]
```diff
@@ -46,15 +58,15 @@ static FixedDueDateSchedules from(JsonObject representation) {
 
   private Predicate<? super JsonObject> isWithin(DateTime date) {
     return schedule -> {
-      DateTime from = DateTime.parse(schedule.getString("from"));
-      DateTime to = DateTime.parse(schedule.getString("to"));
+      DateTime from = DateTime.parse(schedule.getString("from")).withZoneRetainFields(timeZone);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: This appears to convert the timezone, without changing the field values, is that correct?

Can you help me understand why this is needed? 

Does this not make these values potentially incorrect? If the schedule has a `from` of 05:00 because it is meant to be midnight in a non-UTC timezone, won't this make this 05:00 in the new timezone?

File: src/main/java/org/folio/circulation/support/JsonPropertyWriter.java (hunk 3 above)
Comment (OleksiiMaksymov) [reply src/main/java/org/folio/circulation/support/JsonPropertyWriter.java]: Yes, when testing the time zone, it turned out that the `due date` value in the response to "checkout" is not the UTC. Date saves the applied zone.

File: descriptors/ModuleDescriptor-template.json (hunk 4 above)
Comment (OleksiiMaksymov) [reply descriptors/ModuleDescriptor-template.json]: There are no changes in the interface definition, but additional calls (mod-config) have been added to API for `check-out` and `renewal`

File: src/main/java/org/folio/circulation/support/JsonPropertyWriter.java (hunk 3 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/support/JsonPropertyWriter.java]: Ok, I think we may need to revisit the decision to do this conversion here. I would’ve thought that without this change, the code would use the timezone in the date time provided, which should be UTC

File: descriptors/ModuleDescriptor-template.json (hunk 4 above)
Comment (marcjohnson-kint) [reply descriptors/ModuleDescriptor-template.json]: The additional calls are IMO implementation aspects of this change, rather than interface. 

I’m aware that the module descriptor is a grey area, as it binds an interface to an implementation and describes endpoint permissions.

To me, the permissions granted are part of the binding and not the interface.

Does that make sense?

If there are no interface changes, please revert this incrementing of the interface version here and in the RAML.

Comment (OleksiiMaksymov) [reply descriptors/ModuleDescriptor-template.json]: ok

File: descriptors/ModuleDescriptor-template.json (hunk 5 above)
Comment (OleksiiMaksymov) [reply descriptors/ModuleDescriptor-template.json]: done

File: descriptors/ModuleDescriptor-template.json (hunk 4 above)
Comment (OleksiiMaksymov) [reply descriptors/ModuleDescriptor-template.json]: done

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java (hunk 7 above)
Comment (OleksiiMaksymov) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: As far as I understood from the code, this method is called in "LoanPolicy" only for tests.
And at the moment it makes sense to hardcode.

File: src/main/java/org/folio/circulation/domain/LoanRenewalService.java (hunk 6 above)
Comment (OleksiiMaksymov) [reply src/main/java/org/folio/circulation/domain/LoanRenewalService.java]: yes, it is not yet supported for overridden renewal.
In the current implementation, we have provided the ability to change the time zone.
Perhaps in the future, this will be taken into account.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/LoanRenewalService.java]: Is the scope for supporting timezone in renewal overrides in this story, or a different story, as part of this feature?

If you don't know, that's ok, I will ask on the story in JIRA.

File: src/test/java/api/support/builders/TimeZoneConfigBuilder.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
+package api.support.builders;
+
+import io.vertx.core.json.JsonObject;
+
+public class TimeZoneConfigBuilder extends JsonBuilder implements Builder {
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/builders/TimeZoneConfigBuilder.java]: Is there anything that makes this specific to time zone configuration?

It seems to me this only reflects the general structure of a configuration record?

File: src/test/java/api/support/builders/TimeZoneConfigBuilder.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  public JsonObject create() {
+    return this.representation;
+  }
+
+  @Override
+  public String toString() {
+    return this.representation.toString();
+  }
+
+  JsonObject toJson() {
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/builders/TimeZoneConfigBuilder.java]: What is the difference between this and `create`?

File: src/test/java/org/folio/circulation/domain/ConfigurationServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  private static ConfigurationService service;
+
+  @BeforeClass
+  public static void before() {
+    service = new ConfigurationService();
+  }
+
+  @Test
+  public void testUtcTimeZone() {
+    String timeZoneValue = "{\"locale\":\"en-US\",\"timezone\":\"UTC\"}";
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/domain/ConfigurationServiceTest.java]: I'd preferred that JSON wasn't created in hard coded strings, as I find this is brittle to changes, and repetitive.

Personally, I'd use something similar to the mechanism I proposed in the ConfigurationExample class.

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java (hunk 7 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: Why not just use the `static FixedDueDateSchedules from(JsonObject representation, DateTimeZone timeZone)` method and pass UTC in from the tests?

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java (hunk 8 above)
Comment (OleksiiMaksymov) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: Comparing the due date in the interval of a fixed schedule should take into account the selected time zone since a time zone has already been added to the due date.

File: ramls/circulation.raml [hunk 12]
```diff
@@ -1,6 +1,6 @@
 #%RAML 1.0
 title: Circulation
-version: v7.4
+version: v7.5
```
Comment (marcjohnson-kint) [inline ramls/circulation.raml]: This change also needs removing, based upon our conversation about the interface version not changing

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java (hunk 8 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: Hmm, the due dates for fixed due date schedules are specified as date and time already, and therefore should already have timezone information.

If so, IMO it is the responsibility of the client specifying the fixed due date schedules to ensure they are correctly respecting the time zone.

I don't think we should be overriding that decision here. Have I missed something?

@skoczko @kostyantyn-kh What do you think?

File: ramls/circulation.raml (hunk 12 above)
Comment (OleksiiMaksymov) [reply ramls/circulation.raml]: done

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java (hunk 8 above)
Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: @marc-johnson It appears that fixed due date schedule effectively stores date-only values. Due to the fact that RMB only has a date-time format, it adds zeroes for time and time zone offset.
My vision is that fixed due date schedules can work as calender do: store date-only values, that are converted to date-time by a client (that is aware of time zone).
Do you have concerns about this approach?

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: According to the [schema)(
https://github.com/folio-org/mod-circulation-storage/blob/5d3caf5a7bb69afec62aedfac2c61733f0d6cbd2/ramls/schedule.json#L6) the schedules define dates and times.

In my opinion, when dates and times are specified with a timezone, they should be respected.

If what the UI / sample records defines for fixed due date schedules is incorrect, now that we have started working with time zones, I think that should be fixed, and not compensated for in the logic in circulation.

If, alternatively, we want to convert the schedules to be dates only, like calendar, then we should plan that change in and explicitly model it that way.

Maybe we need to move this conversation to JIRA so we can include the product owners.

File: src/test/java/org/folio/circulation/domain/ConfigurationServiceTest.java (hunk 11 above)
Comment (OleksiiMaksymov) [reply src/test/java/org/folio/circulation/domain/ConfigurationServiceTest.java]: Fixed

File: src/test/java/api/support/builders/TimeZoneConfigBuilder.java (hunk 9 above)
Comment (OleksiiMaksymov) [reply src/test/java/api/support/builders/TimeZoneConfigBuilder.java]: Yes, that's right, this class keeps the record structure for configurations.
Fixed

File: src/test/java/api/support/builders/TimeZoneConfigBuilder.java (hunk 10 above)
Comment (OleksiiMaksymov) [reply src/test/java/api/support/builders/TimeZoneConfigBuilder.java]: Fixed

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java (hunk 8 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: @dmytrotur @OleksiiMaksymov Please can you write this up on the JIRA issue for CIRC-238, I think we need to discuss this with a broader audience

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: @marcjohnson-kint Ok, I'll do.

Comment (dmytrotur) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: @marc-johnson Not raised that question yet. It seems like it can take some time to discuss this. Can we revert this change and merge the pull request for now, and discuss it later due to the Q1?

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: @dmytrotur Sure, I'm happy if this is removed whilst we discuss it further.

And if this surfaces during acceptance testing we can discuss the best resolution on a case by case basis.

File: src/main/java/org/folio/circulation/support/JsonPropertyWriter.java (hunk 3 above)
Comment (OleksiiMaksymov) [reply src/main/java/org/folio/circulation/support/JsonPropertyWriter.java]: I think this should be converted on the domain side. I'll revert the changes.

File: src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java (hunk 8 above)
Comment (OleksiiMaksymov) [reply src/main/java/org/folio/circulation/domain/policy/FixedDueDateSchedules.java]: Revert changes

File: src/main/java/org/folio/circulation/domain/LoanRenewalService.java (hunk 6 above)
Comment (OleksiiMaksymov) [reply src/main/java/org/folio/circulation/domain/LoanRenewalService.java]: Revert changes

File: src/main/java/org/folio/circulation/support/JsonPropertyWriter.java (hunk 3 above)
Comment (OleksiiMaksymov) [reply src/main/java/org/folio/circulation/support/JsonPropertyWriter.java]: Do you agree that I will revert the changes applied to `write()`?

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/support/JsonPropertyWriter.java]: Yes, if the domain converts the relevant date and times to UTC, then please revert this chsnge to write

Comment (OleksiiMaksymov) [reply src/main/java/org/folio/circulation/support/JsonPropertyWriter.java]: done
